redis/src/cluster_async/mod.rs

```diff
@@ -499,52 +529,209 @@ where
+        //
+        let (_, random_conn) =
+            get_random_connection(&connections_and_metadata.connections, None).await; // TODO Only do this lookup if the first check fails
+        let connection_future = {
+            let addr = addr.clone();
+            let connections_and_metadata = connections_and_metadata.clone();
+            async move {
+                match connect_and_check(&addr, connections_and_metadata.cluster_params.clone())
+                    .await
+                {
+                    Ok(conn) => conn,
+                    Err(_) => random_conn.await,
                 }
             }
-            .boxed()
-            .shared();
-            self.connections
-                .insert(addr.clone(), connection_future.clone());
-            (addr, connection_future)
+        }
+        .boxed()
+        .shared();
+        connections_and_metadata
+            .connections
+            .lock()
+            .await
+            .insert(addr.clone(), connection_future.clone());
+        (addr, connection_future)
+    }
+
+    fn add_future(
+        connections_and_metadata: Arc<ConnectionsAndMetadata<C>>,
+        func: fn(C, Arc<Cmd>) -> RedisFuture<'static, Response>,
+        cmd: Arc<Cmd>,
+        futures: &mut Vec<RedisFuture<Response>>,
+        slot: &SlotAddrs,
+        slot_addr: &SlotAddr,
+    ) {
+        let addr = slot.slot_addr(slot_addr).to_string();
+        futures.push(
+            (async move {
+                let (_, conn) =
+                    Self::get_connection_from_parts(addr, connections_and_metadata).await;
+                let conn = conn.await;
+                func(conn, cmd.clone()).await
+            })
+            .boxed(),
+        );
+    }
+
+    async fn execute_on_all_nodes(
+        func: fn(C, Arc<Cmd>) -> RedisFuture<'static, Response>,
+        cmd: &Arc<Cmd>,
+        only_primaries: bool,
+        connections_and_metadata: Arc<ConnectionsAndMetadata<C>>,
+    ) -> (OperationTarget, RedisResult<Response>) {
+        let operation_count = if only_primaries {
+            connections_and_metadata.connection_count
         } else {
-            // Return a random connection
-            get_random_connection(&self.connections, None)
+            connections_and_metadata.connection_count * 2
+        };
+
+        let mut futures = Vec::with_capacity(operation_count);
+        for slot in connections_and_metadata.slots.values() {
+            Self::add_future(
+                connections_and_metadata.clone(),
+                func,
+                cmd.clone(),
+                &mut futures,
+                slot,
+                &SlotAddr::Master,
+            );
+
+            if !only_primaries {
+                Self::add_future(
+                    connections_and_metadata.clone(),
+                    func,
+                    cmd.clone(),
+                    &mut futures,
+                    slot,
+                    &SlotAddr::Replica,
```

Comment:
here - I think that for clusters without replicas we'll send the command twice to the primary

Reply:
same